### PR TITLE
feat(ticket#99):Header Component with Title

### PR DESCRIPTION
**Pull request by nadirhussain786**
#### Summary:
This PR introduces a reusable header component that includes breadcrumbs, a search input, and a create button. The component is built using Shadcn UI in a Next.js project with TypeScript. It also includes a responsive design, leveraging a custom media query hook to adjust button size based on screen width.

#### Changes:
- Added a `Header` component with:
  - Breadcrumbs for easy navigation
  - A search input field
  - A create button
- Implemented a custom `useMediaQuery` hook for responsive design.
  
#### Technologies Used:
- Next.js (with TypeScript)
- Shadcn UI components
- Custom media query hook for responsiveness

image

image

### CODE REVIEW

File: src/hooks/use-media-query.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  sm: "(max-width: 640px)",
+  md: "(max-width: 768px)",
+  lg: "(max-width: 1024px)",
+  xl: "(max-width: 1280px)",
+};
+
+const useMediaQuery = (query: keyof typeof BREAK_POINTS): boolean => {
+  const [matches, setMatches] = useState<boolean>(false);
+
+  useEffect(() => {
```
Comment (muhammad-bilal-ali-khokhar) [inline src/hooks/use-media-query.ts:13]: @nadirhussain786  I'll suggest you useMemo and you can impove this code

**Review by muhammad-bilal-ali-khokhar — COMMENTED** [review]
remove commented code

**Review by muhammad-bilal-ali-khokhar — COMMENTED** [review]
@nadirhussain786  UI looks fantastic on both mobile and desktop 🔥

File: src/app/member/components/MemeberHeader.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+  return (
+    <div className="sticky top-0 z-50 w-full">
+      {isSidebarOpen && (
+        <Sidebar isSidebarOpen={isSidebarOpen} toggleSidebar={toggleSidebar} />
+      )}
+      <header className="bg-white shadow-sm sticky top-0">
+        <div className="flex items-center justify-between px-4 py-4">
+          <div className="flex items-center">
+            {pathname !== "/login" && (
```
[the remark this answers is not in the document]
Comment (nadirhussain786) [reply src/app/member/components/MemeberHeader.tsx:30]: fixed

File: src/app/member/components/MemeberHeader.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            )}
+            <h1
+              className={`${
+                !isTilleHide ? "block" : "hidden"
+              } text-sm font-semibold text-gray-800 truncate md:block`}
+            >
+              <MemberBreadCrumbs />
+            </h1>
+          </div>
+          {pathname !== "/login" && (
```
[the remark this answers is not in the document]
Comment (nadirhussain786) [reply src/app/member/components/MemeberHeader.tsx:48]: fixed

File: src/components/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+vi.mock("next/navigation", () => ({
+  usePathname: vi.fn(),
+}));
+
+describe("BasedBreadCrumb", () => {
+  it("renders the correct breadcrumb links", () => {
+    (usePathname as Mock).mockReturnValue("/member/add");
+
+    // render(<BasedBreadCrumb breadcrumbs={mockBreadcrumbs} />);
```
[the remark this answers is not in the document]
Comment (nadirhussain786) [reply src/components/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx:20]: fixed

File: src/app/member/components/MemeberHeader.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
+"use client";
+import { Sidebar } from "@/app/layout/components/Sidebar";
+import { Button } from "@/components/ui/button";
+import { Input } from "@/components/ui/input";
+import { HamburgerMenuIcon, MagnifyingGlassIcon } from "@radix-ui/react-icons";
+import { usePathname } from "next/navigation";
+import React, { useState } from "react";
+import { MemberBreadCrumbs } from "./MemberBreadCrumbs";
+import useMediaQuery from "@/hooks/use-media-query";
+import { Plus } from "lucide-react";
```
Comment (Mubashar1009) [inline src/app/member/components/MemeberHeader.tsx:10]: Use prefix `Icon` PlusIcon

File: src/app/member/components/MemeberHeader.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
+"use client";
+import { Sidebar } from "@/app/layout/components/Sidebar";
+import { Button } from "@/components/ui/button";
+import { Input } from "@/components/ui/input";
+import { HamburgerMenuIcon, MagnifyingGlassIcon } from "@radix-ui/react-icons";
+import { usePathname } from "next/navigation";
+import React, { useState } from "react";
```
Comment (Mubashar1009) [inline src/app/member/components/MemeberHeader.tsx:7]: We should use react import on the top

File: src/app/member/components/MemeberHeader.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import { MemberBreadCrumbs } from "./MemberBreadCrumbs";
+import useMediaQuery from "@/hooks/use-media-query";
+import { Plus } from "lucide-react";
+import { Routes } from "@/lib/routes";
+export const MemeberHeader = () => {
```
Comment (Mubashar1009) [inline src/app/member/components/MemeberHeader.tsx:12]: we should keeps space between imports and compoennt

File: src/app/member/components/MemeberHeader.tsx [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+                size="icon"
+                onClick={toggleSidebar}
+                className="mr-2 lg:hidden"
+              >
+                <HamburgerMenuIcon className="h-6 w-6 text-black" />
+              </Button>
+            )}
+            <h1
+              className={`${
+                !isTitleHide ? "block" : "hidden"
```
Comment (Mubashar1009) [inline src/app/member/components/MemeberHeader.tsx:43]: why we are using component in `h1`

File: src/app/member/components/MemeberHeader.tsx [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+                    setIsTitleHide(false);
+                  }}
+                  className={`pr-4 py-2 h-7 ${
+                    isTitleHide ? "w-32 pl-10" : "w-2 pl-6"
+                  } md:w-60 md:pl-10 rounded text-xs`}
+                />
+                <MagnifyingGlassIcon className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-600 w-4 h-4" />
+              </div>
+              <Button data-testid="addMemberButton" className="h-7">
+                <Plus /> {!isMediumScreen ? "Add Member" : null}
```
Comment (Mubashar1009) [inline src/app/member/components/MemeberHeader.tsx:69]: We should use `&&` instead of condition

File: src/app/member/page.tsx [hunk 10]
```diff
@@ -1,9 +1,14 @@
 import React from "react";
 import { MemberList } from "./components/MemberList";
+import { MemeberHeader } from "./components/MemeberHeader";
+
 const Page = () => {
   return (
-    <div className="p-8">
-      <MemberList />
+    <div>
```
Comment (Mubashar1009) [inline src/app/member/page.tsx:7]: Can we use React Fragment instead of div

File: src/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      removeListener: vi.fn(),
+      addEventListener: vi.fn(),
+      removeEventListener: vi.fn(),
+    })),
+  });
+});
+
+describe("BasedBreadCrumb", () => {
+  it("renders the correct breadcrumb links", () => {
+    (usePathname as Mock).mockReturnValue("/member/add");
```
Comment (Mubashar1009) [inline src/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx:34]: use enum here

File: src/hooks/UseMediaQuery.test.tsx [hunk 12]
```diff
@@ -0,0 +1,26 @@
+import { expect, it, describe } from "vitest";
+import { renderHook } from "@testing-library/react-hooks";
+import useMediaQuery from "./use-media-query";
+
+describe("useMediaQuery", () => {
+  beforeAll(() => {
```
Comment (Mubashar1009) [inline src/hooks/UseMediaQuery.test.tsx:6]: I think here we should not need this

File: src/app/member/components/MemberHeader.test.tsx [hunk 13]
```diff
@@ -0,0 +1,30 @@
+import { render, screen } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+import { it, describe, vi } from "vitest";
+import { MemeberHeader } from "./MemeberHeader";
+
+describe("Header Component", () => {
+  beforeAll(() => {
```
Comment (Mubashar1009) [inline src/app/member/components/MemberHeader.test.tsx:7]: I think we should not need these props here

File: src/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  { href: Routes.Member, label: "Member" },
+  { href: Routes.AddMember, label: "New Member" },
+];
+
+vi.mock("next/navigation", () => ({
+  usePathname: vi.fn(),
+}));
+
+beforeAll(() => {
+  Object.defineProperty(window, "matchMedia", {
```
Comment (Mubashar1009) [inline src/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx:18]: here also

File: src/app/member/components/MemberBreadCrumbs.tsx [hunk 15]
```diff
@@ -0,0 +1,17 @@
+import { BasedBreadCrumb } from "@/common/BasedBreadCrumb/BasedBreadCrumb";
```
[the remark this answers is not in the document]
Comment (nadirhussain786) [reply src/app/member/components/MemberBreadCrumbs.tsx:1]: @Mubashar1009 We have a shared breadcrumb component called `BasedBreadCrumb` located in the common folder.

File: src/app/member/components/MemberBreadCrumbs.tsx [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+import { Routes } from "@/lib/routes";
+import { BreadcrumbsTypes } from "@/types";
+import React from "react";
```
[the remark this answers is not in the document]
Comment (nadirhussain786) [reply src/app/member/components/MemberBreadCrumbs.tsx:4]: fixed

File: src/app/member/components/MemberBreadCrumbs.tsx [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+
+const breadcrumbs: BreadcrumbsTypes[] = [
+  { href: Routes.Member, label: "Member" },
+  { href: Routes.AddMember, label: "New Member" },
+];
+
+export const MemberBreadCrumbs = () => {
+  return (
+    <div>
+      <BasedBreadCrumb breadcrumbs={breadcrumbs} />
```
[the remark this answers is not in the document]
Comment (nadirhussain786) [reply src/app/member/components/MemberBreadCrumbs.tsx:14]: @Mubashar1009 I used the <div> in the MemberBreadCrumbs component to act as a container for the BasedBreadCrumb. Here are a couple of reasons for including it:

- Structural Purpose
- Future Expansion

If we don't require these benefits, we can remove the <div> and return BasedBreadCrumb directly.

File: src/app/member/page.tsx (hunk 10 above)
Comment (nadirhussain786) [reply src/app/member/page.tsx:7]: @Mubashar1009  Nope for styling we can’t use `React Fragment`.

File: src/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx (hunk 11 above)
Comment (nadirhussain786) [reply src/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx:34]: Fixed

File: src/hooks/UseMediaQuery.test.tsx (hunk 12 above)
Comment (nadirhussain786) [reply src/hooks/UseMediaQuery.test.tsx:6]: @Mubashar1009  The `beforeAll` block is currently used to mock the matchMedia function, which allows us to simulate different screen sizes in our tests."

File: src/app/member/components/MemberHeader.test.tsx (hunk 13 above)
Comment (nadirhussain786) [reply src/app/member/components/MemberHeader.test.tsx:7]: @Mubashar1009 same here

File: src/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx (hunk 14 above)
Comment (nadirhussain786) [reply src/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx:18]: @Mubashar1009  same here

File: src/app/member/components/MemberHeader.test.tsx [hunk 18]
```diff
@@ -0,0 +1,30 @@
+import { render, screen } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
```
[the remark this answers is not in the document]
Comment (nadirhussain786) [reply src/app/member/components/MemberHeader.test.tsx:2]: fixed

File: src/app/member/components/MemberHeader.test.tsx [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+        addListener: vi.fn(),
+        removeListener: vi.fn(),
+        addEventListener: vi.fn(),
+        removeEventListener: vi.fn(),
+      })),
+    });
+  });
+  it("Input Event Triggered", async () => {
+    render(<MemeberHeader />);
+    userEvent.click(screen.getByTestId("searchInput"));
```
[the remark this answers is not in the document]
Comment (nadirhussain786) [reply src/app/member/components/MemberHeader.test.tsx:23]: Fixed

File: src/hooks/use-media-query.ts (hunk 1 above)
Comment (nadirhussain786) [reply src/hooks/use-media-query.ts:14]: fixed

File: src/types/index.tsx [hunk 20]
```diff
@@ -9,7 +9,7 @@ export interface Member {
   status: UserStatus;
 }
 
-export interface BreadcrumbLinkType {
+export interface BreadcrumbsTypes {
```
Comment (muhammad-bilal-ali-khokhar) [inline src/types/index.tsx:12]: @nadirhussain786  name is not suitable for `interface`  , or you can export as type instead of `interface`

**Review by muhammad-bilal-ali-khokhar — APPROVED** [review]

File: src/app/member/components/MemberHeader.test.tsx [hunk 21]
```diff
@@ -0,0 +1,40 @@
+import { fireEvent, render, screen } from "@testing-library/react";
+import { it, describe, vi } from "vitest";
+import { MemeberHeader } from "./MemeberHeader";
+
+describe("Header Component", () => {
+  beforeAll(() => {
+    Object.defineProperty(window, "matchMedia", {
```
Comment (Monfernape) [inline src/app/member/components/MemberHeader.test.tsx:7]: What's going on here?

File: src/app/member/components/MemeberHeader.tsx [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+import { HamburgerMenuIcon, MagnifyingGlassIcon } from "@radix-ui/react-icons";
+import { usePathname } from "next/navigation";
+import { MemberBreadCrumbs } from "./MemberBreadCrumbs";
+import useMediaQuery from "@/hooks/use-media-query";
+import { Plus } from "lucide-react";
+import { Routes } from "@/lib/routes";
+
+export const MemeberHeader = () => {
+  const pathname = usePathname();
+  const isMediumScreen = useMediaQuery("md");
```
Comment (Monfernape) [inline src/app/member/components/MemeberHeader.tsx:16]: Can't we just rely on CSS for this? Why use JS?

File: src/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+import { Breadcrumbs } from "@/types";
+import { Routes } from "@/lib/routes";
+
+const mockBreadcrumbs: Breadcrumbs[] = [
+  { href: Routes.Member, label: "Member" },
+  { href: Routes.AddMember, label: "New Member" },
+];
+
+vi.mock("next/navigation", () => ({
+  usePathname: vi.fn(),
```
Comment (Monfernape) [inline src/common/BasedBreadCrumb/BasedBreadCrumb.test.tsx:14]: Nice

File: src/common/BasedBreadCrumb/BasedBreadCrumb.tsx [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+  const isSmallScreen = useMediaQuery("sm");
+
+  const updatedBreadCrumbs = useMemo(() => {
+    const currenLinkIndex = breadcrumbs.findIndex((x) => x.href === pathname);
+    let updatedLinks = breadcrumbs.slice(0, currenLinkIndex + 1);
+    if (isSmallScreen) {
+      updatedLinks = breadcrumbs.slice(-1);
+    }
+    return updatedLinks;
+  }, [pathname]);
```
Comment (Monfernape) [inline src/common/BasedBreadCrumb/BasedBreadCrumb.tsx:30]: This seems wrong. We are missing dependencies from `breadcrumbs` and `isSmallScreen`. How is this working?

**Review by Monfernape — APPROVED** [review]
Looks good, few comments to address

**Pull request merged**